internal/workflow/batch.go

```diff
@@ -85,3 +114,92 @@ func (w *BatchWorkflow) updateBatch(ctx temporalsdk_workflow.Context, state *bat
 
 	return nil
 }
+
+func (w *BatchWorkflow) startSIPWorkflow(
+	ctx temporalsdk_workflow.Context,
+	state *batchWorkflowState,
+	index int,
+	key string,
+	sourceID uuid.UUID,
+) error {
+	// Generate SIP UUID using SideEffect to ensure determinism.
+	var sipUUID uuid.UUID
+	genUUID := temporalsdk_workflow.SideEffect(ctx, func(ctx temporalsdk_workflow.Context) any {
+		return uuid.Must(uuid.NewRandomFromReader(w.rng))
+	})
+	if err := genUUID.Get(&sipUUID); err != nil {
+		return fmt.Errorf("generate SIP UUID: %v", err)
+	}
+
+	// Create SIP.
+	sip := datatypes.SIP{
+		UUID:     sipUUID,
+		Name:     key,
+		Status:   enums.SIPStatusQueued,
+		Batch:    &state.batch,
+		Uploader: state.batch.Uploader,
+	}
+	activityOpts := withLocalActivityOpts(ctx)
+	err := temporalsdk_workflow.ExecuteLocalActivity(
+		activityOpts,
+		createSIPLocalActivity,
+		w.ingestsvc,
+		&createSIPLocalActivityParams{SIP: sip},
+	).Get(activityOpts, nil)
+	if err != nil {
+		return fmt.Errorf("create SIP: %v", err)
+	}
+
+	// Start processing workflow for the SIP, keeping track of the workflow future and execution.
+	var we temporalsdk_workflow.Execution
+	processingCtx := temporalsdk_workflow.WithChildOptions(ctx, temporalsdk_workflow.ChildWorkflowOptions{
+		Namespace:         w.cfg.Temporal.Namespace,
+		TaskQueue:         w.cfg.Temporal.TaskQueue,
+		WorkflowID:        fmt.Sprintf("%s-%s", ingest.ProcessingWorkflowName, sipUUID.String()),
+		ParentClosePolicy: temporalapi_enums.PARENT_CLOSE_POLICY_TERMINATE,
+	})
+	wf := temporalsdk_workflow.ExecuteChildWorkflow(
+		processingCtx,
+		ingest.ProcessingWorkflowName,
+		&ingest.ProcessingWorkflowRequest{
+			SIPUUID:         sipUUID,
+			SIPName:         key,
+			Key:             key,
+			SIPSourceID:     sourceID,
+			Type:            enums.WorkflowTypeCreateAip,
+			RetentionPeriod: -1 * time.Second,
+			// Don't pass BatchUUID yet to avoid waiting for validation.
+			// BatchUUID:       state.batch.UUID,
+		},
+	)
+	err = wf.GetChildWorkflowExecution().Get(processingCtx, &we)
+	if err != nil {
+		return fmt.Errorf("processing workflow: %v", err)
+	}
+
+	// Store SIP details in the batch workflow state.
+	state.addSIPDetails(index, sip, wf, we)
+
+	return nil
+}
+
+func (w *BatchWorkflow) waitForWorkflowsCompletion(ctx temporalsdk_workflow.Context, state *batchWorkflowState) error {
+	selector := temporalsdk_workflow.NewSelector(ctx)
+	for _, sd := range state.sipDetails {
+		selector.AddFuture(sd.workflowFuture, func(f temporalsdk_workflow.Future) {
+			// Ignore error and result, we just want to know when it's done.
+			_ = f.Get(ctx, nil)
+			state.sipsCompleted++
+		})
+	}
+
+	// Wait for all SIP workflows to complete or context cancellation.
+	for !state.allSIPsCompleted() {
```

Comment:
I think you can just do `for range len(state.sipDetails) {` rather than checking `state.allSIPsCompleted()` on each iteration, since selected future always increments `sipsCompleted`.